cached-http-filesystem/src/main/java/azkaban/cachedhttpfilesystem/CachedHttpFileSystem.java

```diff
@@ -0,0 +1,297 @@
+    }
+  }
+
+  // If the base URIs don't have a trailing forward slash the resolving and relativization can get messed up.
+  private static String addTrailingForwardSlash(String in) {
+    return in.endsWith("/") ? in : in + "/";
+  }
+
+  @Override
+  public String getScheme() {
+    return "chttp";
+  }
+
+  @Override
+  public URI getUri() {
+    return uri;
+  }
+
+  @Override
+  public FSDataInputStream open(Path path, int bufferSize) throws IOException {
+    URI relativeURI = this.uri.relativize(path.toUri());
+    if (relativeURI.isAbsolute()) {
+      throw new IOException("Path must be relative or have same prefix as root origin URI.");
+    }
+
+    URI resolvedOriginURI = this.rootOriginURI.resolve(relativeURI);
+    if (!cachingEnabled) {
+      // If caching is disabled, just return from the origin.
+      return downloadFromOrigin(resolvedOriginURI);
+    }
+
+    Path resolvedCachePath = new Path(this.rootCachedURI.resolve(relativeURI));
+    try {
+      // Try to pull from cache
+      return this.cacheFS.open(resolvedCachePath, bufferSize);
+    } catch (FileNotFoundException e) {
+      log.info("Cache miss, file not found: " + resolvedCachePath.toString());
+      // Cache miss, let's download from the origin
+      FSDataInputStream originInputStream = downloadFromOrigin(resolvedOriginURI);
+
+      // Let's persist to the cache
+      Path folderInCacheForFile = resolvedCachePath.getParent();
+      Path tempCacheFile = new Path(folderInCacheForFile, String.format(CACHE_TMP_FILE_TEMPLATE, RAND.nextInt()));
+      FSDataOutputStream outStreamToTmpFile;
+      try {
+        this.cacheFS.mkdirs(folderInCacheForFile);
+        outStreamToTmpFile = this.cacheFS.create(tempCacheFile, false);
+      } catch (IOException e2) {
+        // We failed to create the output stream, so just return the stream from the origin
+        log.warn("Failed to persist file to cache, returning stream from origin: " +
+            resolvedOriginURI.toString(), e2);
+        return originInputStream;
+      }
+
+      // Copy from origin to tmp file in cache
+      IOUtils.copy(originInputStream, outStreamToTmpFile);
+      originInputStream.close();
+      outStreamToTmpFile.close();
+
+      try {
+        // Rename the temporary file to the final file name
+        this.cacheFS.rename(tempCacheFile, resolvedCachePath);
+      } catch (FileAlreadyExistsException e2) {
+        log.info("Another process already persisted this file: " + resolvedCachePath.toString());
+        // Another process beat us to the race - no problem though, that means the file already exists so we
+        // can just swallow this error and return the stream like usual!
+      }
+
+      // Return input stream from cache
+      return this.cacheFS.open(resolvedCachePath, bufferSize);
+    }
+  }
+
+  @Override
+  public FSDataOutputStream create(Path path, FsPermission fsPermission,
+      boolean b, int i, short i1, long l,
```

Comment:
short s

Reply:
This was copied directly from: https://github.com/apache/hadoop/blob/trunk/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/http/AbstractHttpFileSystem.java

Going to leave as is.